### PR TITLE
Update docs

### PR DESCRIPTION
**Pull request by frostney**

### CODE REVIEW

File: docs/bytecode-vm.md [hunk 1]
```diff
@@ -137,6 +137,88 @@ The profiler follows the same singleton-tracker pattern as coverage (`Goccia.Cov
...
+
+Stack-based VMs (like the JVM and WASM) are simpler to compile to and have smaller instruction encoding. Register-based VMs (like Lua 5, LuaJIT, and Dalvik) need fewer instructions per operation and avoid redundant stack manipulations. Register-based was chosen for execution performance.
+
+### Why Two Tiers?
+
+The solution is a split opcode space:
+
+- **Core range (0–127):** register, control-flow, closure, literal, and other hot/stable VM operations.
+- **Semantic range (128–255):** colder language-level operations that are still explicit bytecode, such as generic arithmetic/comparison, imports/exports, and await.
+
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/bytecode-vm.md]: Fixed in 7adeb957. Updated to three-tier model: core (0–127), non-core generic (128–166), semantic helper (167–255 starting at `OP_SEM_FIRST = 167`). Also renamed the heading from "Why Two Tiers?" to "Why Three Tiers?".

File: docs/bytecode-vm.md [hunk 2]
```diff
@@ -137,6 +137,88 @@ The profiler follows the same singleton-tracker pattern as coverage (`Goccia.Cov
...
+This split keeps the dispatch surface organized while still allowing the backend to be explicitly Goccia-specific.
+
+### Why Shared Runtime Values?
+
+The current VM uses `TGocciaValue` directly instead of maintaining a second value representation.
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/bytecode-vm.md]: Fixed in 7adeb957. Rewrote to distinguish `TGocciaRegister` (tagged variant record in the register file, keeping scalars unboxed) from `TGocciaValue` (the shared runtime object model). Values only cross into `TGocciaValue` when leaving the register file at object/runtime boundaries.

File: docs/decision-log.md [hunk 3]
```diff
@@ -1,64 +1,60 @@
...
-## Executive Summary
+## How to add an entry
 
-- **Bytecode fold-in** — The bytecode backend is now Goccia-owned, operating directly on `TGocciaValue` with dedicated opcodes
-- **Evaluator design** — Pure functions, virtual dispatch, `TGocciaControlFlow` records (not exceptions) for break/return
-- **Data structures** — Custom hash maps (4-6x faster than `TDictionary`), `TStringBuffer` (not `TStringBuilder`), no string interning
-- **FPC findings** — Generics have zero runtime cost; virtual dispatch is constant-time; ObjFPC vs Delphi mode has zero performance impact
+```markdown
+**YYYY-MM-DD** · `area` — Summary of the decision and why. [#PR](../../pull/N). [detail link](file.md#anchor).
+```
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/decision-log.md:9]: Fixed in 7adeb957. All 6 occurrences of `../../pull/N` replaced with absolute GitHub URLs (`https://github.com/frostney/GocciaScript/pull/N`).

File: docs/interpreter.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+The evaluator threads state through a `TGocciaEvaluationContext` record rather than using instance variables or globals:
+
+```pascal
+TGocciaEvaluationContext = record
+  Scope: TGocciaScope;
+  OnError: TGocciaThrowErrorCallback;
+  LoadModule: TLoadModuleCallback;
+end;
+```
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/interpreter.md:40]: Fixed in 7adeb957. Added the three missing fields to match `units/Goccia.Evaluator.Context.pas`: `CurrentFilePath: string`, `CoverageEnabled: Boolean`, `DisposalTracker: TObject`.

File: docs/language.md [hunk 5]
```diff
@@ -1,18 +1,79 @@
...
+| [Decorator Metadata](https://github.com/tc39/proposal-decorator-metadata) | 3 | Supported — `Symbol.metadata` for decorator-attached class metadata with inheritance |
+| [Types as Comments](https://tc39.es/proposal-type-annotations/) | 1 | Supported — TypeScript-style annotations parsed, preserved on AST, enforced in bytecode mode |
+| [Enum Declarations](https://github.com/tc39/proposal-enum) | 0 | Supported — frozen, null-prototype enum objects with `Symbol.iterator` |
+| [Temporal](https://tc39.es/proposal-temporal/) | 3 | Supported — `Temporal.PlainDate`, `Temporal.Duration`, `Temporal.Instant`, etc. |
+| [`Math.clamp`](https://github.com/tc39/proposal-math-clamp) | 3 | Supported |
+| [`Math.sumPrecise`](https://github.com/tc39/proposal-math-sum) | 3 | Supported |
+| [`Map.prototype.getOrInsert`](https://github.com/tc39/proposal-upsert) | 3 | Supported — `getOrInsert` and `getOrInsertComputed` |
+| [`Error.isError`](https://github.com/tc39/proposal-is-error) | 4 | Supported |
+| [`RegExp.escape`](https://github.com/tc39/proposal-regex-escaping) | 3 | Supported |
+
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/language.md]: Already resolved. The tables were extracted to `docs/language-tables.md` in a subsequent commit. The current `language.md` has only `## TC39 Proposal Details` — no duplicate heading.

File: docs/language.md [hunk 6]
```diff
@@ -93,7 +154,7 @@ class C {
 - Static properties.
 - Static blocks (`static { ... }`).
 - Inheritance with `extends` and `super`.
-- Decorators (see [Decorators](#decorators) below).
+- Decorators (see [Decorators](#decorators) in TC39 Proposals below).
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/language.md]: Already resolved. The current `language.md` uses `#decorators-stage-3` at both line 96 and line 434. The link was corrected in a subsequent commit.

File: docs/value-system.md [hunk 7]
```diff
@@ -638,6 +638,98 @@ Self-references in initializers are supported via a child scope that binds each
...
+
+- **Aliased references** — A value assigned to multiple variables, captured in a closure, and stored in an array has no single owner. Determining when to free it requires tracking all references.
+- **Shared prototype singletons** — String, Number, Array, Set, Map, Function, and Symbol prototype objects are class-level singletons shared across all instances of their type. Each type's `InitializePrototype` creates the singleton once (guarded by `if Assigned`) and pins it with `TGarbageCollector.Instance.PinObject`. Manual lifetime tracking of these shared singletons would be fragile.
+- **Closure captures** — Arrow functions capture their enclosing scope, creating non-obvious reference chains between scopes and values.
+
+**Why not reference counting (via `TInterfacedObject`)?**
+
+`TGocciaValue` inherits from `TInterfacedObject`, which provides automatic reference counting. However, values are stored as class references (`TGocciaValue`), not interface references. Switching to interface variables throughout the evaluator would require a large-scale refactor and introduce circular reference issues (objects referencing their prototypes and vice versa).
+
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/value-system.md]: Fixed in 7adeb957. Corrected to `TGocciaValue = class(TGCManagedObject)` where `TGCManagedObject` is a plain `TObject` descendant — no reference counting. The rationale paragraph now accurately explains why interface-based refcounting isn't used.

File: docs/language.md [hunk 8]
```diff
@@ -161,7 +115,7 @@ C[Symbol.metadata].decorated; // true
 - Bitwise operators (`&`, `|`, `^`, `<<`, `>>`, `>>>`).
 - Ternary operator (`? :`).
 - Numeric literals: decimal, hex (`0x`), binary (`0b`), octal (`0o`), scientific notation, and numeric separators (`1_000_000`, `0xFF_FF`, `0b1010_0001`, `0o77_77`, `1.5e1_0`).
-- Template literals with interpolation and tagged templates (``tag`...```, `String.raw`). Tagged templates support the TC39 Template Literal Revision (ES2018): malformed escape sequences (e.g., `\u{`, `\xG1`) set the cooked segment to `undefined` while preserving the raw source text; untagged templates still throw `SyntaxError` for invalid escapes.
+- Template literals with interpolation and tagged templates (``tag`...```,`String.raw`). Tagged templates support the TC39 Template Literal Revision (ES2018): malformed escape sequences (e.g.,`\u{`,`\xG1`) set the cooked segment to`undefined` while preserving the raw source text; untagged templates still throw `SyntaxError` for invalid escapes.
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/language.md:118]: But this landed already in #287 - keep up little 🐰

File: docs/contributing/code-style.md [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+```
+
+Overflow and range checks are **enabled** — correctness is prioritized over raw performance.
+
+### Naming Conventions
+
+| Element | Convention | Example |
+|---------|-----------|---------|
+| Units | `Goccia.<Category>.<Name>.pas` | `Goccia.Values.Primitives.pas` |
+| Classes | `TGoccia<Name>` prefix | `TGocciaObjectValue` |
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/contributing/code-style.md:46]: Updated in 56595365. Corrected the wording — these are shared infrastructure units that are intentionally outside the `Goccia.*` namespace (not legacy units). New phrasing: "Shared infrastructure units that are not GocciaScript-specific (`BaseMap.pas`, `HashMap.pas`, etc.) intentionally live outside the `Goccia.*` namespace. Only project-specific units use the `Goccia.<Category>.<Name>.pas` convention."

File: docs/contributing/tooling.md [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+./format.pas --check
+```
+
+### What It Enforces
+
+All of the following are **auto-fixed** (not just warned about):
+
+- **Uses clauses**: one unit per line, grouped (System > Third-party > Project > Relative), alphabetically sorted within each group, blank line between groups. Units with an `in` path are always in the Relative group; `Goccia.*` units are Project; known FPC standard library units are System; everything else is Third-party.
+- **Function naming**: capitalizes the first letter of function, procedure, constructor, and destructor names to enforce PascalCase. Renames all references within the same file. External C bindings are excluded.
+- **Parameter naming**: adds the `A` prefix to multi-letter parameters (e.g., `Value` -> `AValue`) and renames all references within the function scope (declaration, local variables, and body). Single-letter parameters and Pascal keyword conflicts are skipped.
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/contributing/tooling.md:38]: No change needed. The doc already says "capitalizes the first letter of function, procedure, constructor, and destructor names" which is exactly what `FixFuncNames` does. No overstatement found.

File: docs/interpreter.md [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+  Scope: TGocciaScope;
+  OnError: TGocciaThrowErrorCallback;
+  LoadModule: TLoadModuleCallback;
+  CurrentFilePath: string;
+  CoverageEnabled: Boolean;
+  DisposalTracker: TObject; // TGocciaDisposalTracker or nil
+end;
+```
+
+This keeps evaluator functions pure — all dependencies are explicit parameters. The `OnError` callback is also stored on `TGocciaScope` and propagated to child scopes, so closures always have access to the error handler without global mutable state.
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/interpreter.md:42]: No change needed. `OnError` exists on both `TGocciaScope` (property, line 70 of `Goccia.Scope.pas`) and `TGocciaEvaluationContext` (record field). The scope's `OnError` is propagated via `CreateChild` as described. The doc is accurate.

File: docs/language.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (frostney) [reply docs/language.md:118]: No change needed. The implementation fully supports TC39 Template Literal Revision: the lexer tolerates malformed escapes in tagged templates (sets `CookedValid := False` per segment), while untagged templates still throw `SyntaxError`. The evaluator/VM emit `undefined` for invalid cooked segments. Verified in `Goccia.Lexer.pas`, `Goccia.Parser.pas`, and `Goccia.Evaluator.Expressions.pas`.

File: docs/language.md [hunk 12]
```diff
@@ -490,47 +332,112 @@ for await (const x of asyncIterable) {
...
 
-`with` creates ambiguous scope and is deprecated even in JavaScript's strict mode. Note that `with` is a reserved keyword in GocciaScript (it cannot be used as a variable name), but it can be used as a property name (e.g., `obj.with`).
+Lowercase tags produce string tag names (`"div"`, `"span"`); uppercase tags are passed as identifier references (component functions/classes).
 
-### Labeled Statements
+**Custom factory:** The factory and fragment function names can be overridden per-file using pragma comments (`@jsxFactory`, `@jsxFragment`) at the top of the file, before any code.
 
-**Excluded.** No alternative needed. The parser accepts labeled statements but strips the label and emits a warning:
+The transformer generates an internal source map for accurate error line/column reporting. JSX is enabled by default via `DefaultPreprocessors`; embedders can disable it with `Exclude(Engine.Preprocessors, ppJSX)`.
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/language.md]: Fixed in 8ca28042. Changed `Exclude(Engine.Preprocessors, ppJSX)` to `Engine.Preprocessors := Engine.Preprocessors - [ppJSX]` since `Preprocessors` is a property with a setter.

File: docs/value-system.md [hunk 13]
```diff
@@ -22,7 +22,7 @@ classDiagram
     TGocciaValue <|-- TGocciaStringLiteralValue
     TGocciaValue <|-- TGocciaSymbolValue
     TGocciaValue <|-- TGocciaObjectValue
-    TGocciaValue <|-- TGocciaNativeFunction
+    TGocciaValue <|-- TGocciaNativeFunctionValue
     TGocciaValue <|-- TGocciaError
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/value-system.md]: Fixed in 8ca28042. Corrected the hierarchy diagram: added `TGocciaFunctionBase` as the intermediate class with `TGocciaFunctionValue`, `TGocciaNativeFunctionValue`, and `TGocciaBoundFunctionValue` as children. Also fixed `TGocciaClassValue` inheritance and removed `TGocciaError` (which extends `Exception`, not `TGocciaValue`).

File: README.md [hunk 14]
```diff
@@ -286,85 +213,52 @@ flowchart LR
...
-| Bytecode Units | `Goccia.Bytecode*.pas` | Opcode definitions, templates, modules, debug info, binary I/O |
-| VM | `Goccia.VM*.pas` | Register-based execution on `TGocciaValue` |
-| Backend | `Goccia.Engine.Backend.pas` | Bytecode backend orchestration |
-| Binary I/O | `Goccia.Bytecode.Binary.pas` | `.gbc` file serialization/deserialization |
+Both backends share the same value types, built-ins, scope chain, and mark-and-sweep GC. The bytecode backend is a Goccia-owned VM that executes directly on `TGocciaValue`, not a generic VM layer.
 
-The bytecode backend is a Goccia-owned VM surface with shared runtime objects across interpreter and bytecode mode.
-
-See [Architecture](docs/architecture.md) for the interpreter deep-dive and [Bytecode VM](docs/bytecode-vm.md) for the bytecode VM architecture.
+See [Architecture](docs/architecture.md) for pipelines and layers, [Interpreter](docs/interpreter.md) for the tree-walk backend, [Bytecode VM](docs/bytecode-vm.md) for the bytecode backend, and [Core patterns](docs/core-patterns.md) for implementation patterns and internal terminology.
```
[the remark this answers is not in the document]
Comment (frostney) [reply README.md]: Fixed in 8ca28042. Updated to mention `TGocciaRegister` tagged values with unboxed scalars, falling back to `TGocciaValue` at runtime/object boundaries.

File: docs/contributing/code-style.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (frostney) [reply docs/contributing/code-style.md:46]: Mr. 🐰 , make sure the learning is not legacy unit wording as the comment has been updated. Those are shared infrastructure units.

File: docs/contributing/code-style.md [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+
+Which sets:
+
+```pascal
+{$mode delphi}           // Delphi-compatible mode
+{H+}                     // Long strings by default
+{$overflowchecks on}     // Runtime overflow checking
+{$rangechecks on}        // Runtime range checking
+{$modeswitch advancedrecords}   // Records with methods
+```
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/contributing/code-style.md:37]: No change needed. The actual `Goccia.inc` file uses `{H+}` (without `$`), which is valid FPC syntax — it's the short-form directive for long strings. The doc matches the source file.

File: docs/value-system.md [hunk 16]
```diff
@@ -232,24 +210,17 @@ function BooleanValue(B: Boolean): TGocciaBooleanLiteralValue;
...
 
-**Checking for special values:** Always use the property accessors (`IsNaN`, `IsInfinity`, `IsNegativeZero`) rather than inspecting `FValue` directly. Since special values store `0.0` in `FValue`, standard floating-point checks like `Math.IsNaN(FValue)` will return incorrect results. `IsNegativeZero` uses an endian-neutral `Int64 absolute` overlay (`Bits < 0`) to detect the sign bit — see [docs/code-style.md](code-style.md) for the pattern. When checking for actual zero (not a special value), use `(Value = 0) and not IsNaN and not IsInfinite` — see `IsActualZero` in `Goccia.Evaluator.Arithmetic.pas` for the canonical helper. Similarly, the sort comparator in `Goccia.Values.ArrayValue.pas` uses `NumericRank` to map each special value to a distinct `Double` for correct ordering.
+**Checking for special values:** Use the property accessors (`IsNaN`, `IsInfinity`, `IsNegativeZero`) which delegate to `Math.IsNaN`, `Math.IsInfinite`, and an endian-neutral `Int64 absolute` sign-bit check respectively. See [Tooling](contributing/tooling.md#endian-dependent-byte-indexing) for the endian-neutral pattern.
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/value-system.md:220]: Fixed in fa83c8fc. The design rationale had outdated wording ("dedicated enum") from an earlier version. The actual implementation uses a single `Double` with IEEE 754 property accessors. Aligned the rationale with the Numbers section and removed the contradictory references.

File: docs/value-system.md [hunk 17]
```diff
@@ -535,12 +507,12 @@ All functions share a prototype that provides `call`, `apply`, and `bind`:
...
+### Native Functions (`TGocciaNativeFunctionValue`)
 
 Wraps a Pascal callback for built-in operations:
 
 ```pascal
-TGocciaNativeFunctionCallback = function(
+TGocciaNativeFunctionValueCallback = function(
   Args: TGocciaValueArray;
   ThisValue: TGocciaValue
 ): TGocciaValue;
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/value-system.md]: Already fixed in 56595365. The callback signature was updated to `TGocciaNativeFunctionCallback` with `TGocciaArgumentsCollection` and `of object` — matching `Goccia.Values.NativeFunctionCallback.pas`.

File: docs/value-system.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (frostney) [reply docs/value-system.md:220]: Yes please, remove the old learning and replace it with the updated one

File: docs/value-system.md [hunk 18]
```diff
@@ -638,9 +613,75 @@ Self-references in initializers are supported via a child scope that binds each
...
+- **Callback correctness** — Array prototype methods (`map`, `filter`, `reduce`) pass `undefined` as `ThisValue` to callbacks. Arrow function callbacks correctly inherit their enclosing method's `this`; extracted method references receive `undefined`, preventing accidental `this` leakage.
+
+### Property Descriptor System
+
+Object properties follow ECMAScript's property descriptor model:
+
+- **Data descriptors** — `{ value, writable, enumerable, configurable }`
+- **Accessor descriptors** — `{ get, set, enumerable, configurable }`
+- **Insertion order** — Properties maintain their creation order, matching JavaScript's `Object.keys()` ordering guarantee.
+- **Descriptor merging** — `Object.defineProperty` merges the new descriptor with the existing one when the property already exists. Unspecified attributes retain their current values rather than resetting to defaults. This matches ECMAScript spec behavior (e.g., `Object.defineProperty(obj, "x", { enumerable: false })` only changes `enumerable`, preserving `writable`, `configurable`, and `value`).
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/value-system.md]: Fixed in 313f96d1. Changed both occurrences of "ECMAScript spec behavior" to "ECMAScript specification behavior".

File: docs/value-system.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (frostney) [reply docs/value-system.md]: Follow-up fix in 9c836e4f — removed the remaining `TGocciaError` block from the Mermaid diagram. It extends `Exception`, not `TGocciaValue`, so it shouldn't appear in the value hierarchy.

File: docs/value-system.md [hunk 19]
```diff
@@ -638,9 +610,75 @@ Self-references in initializers are supported via a child scope that binds each
...
+
+**Why property accessors matter:**
+
+- **NaN identity** — IEEE 754 `NaN ≠ NaN`, but JavaScript operations need reliable classification. Always use `IsNaN` rather than raw numeric comparison.
+- **Negative zero** — `-0` and `+0` are equal in IEEE 754 but distinguishable in JavaScript (`Object.is(-0, +0)` is `false`). The `IsNegativeZero` accessor encapsulates the sign-bit check.
+- **Display correctness** — `NaN.toString()` must return `"NaN"`, not a floating-point artifact. The accessors ensure correct string conversion.
+
+**Pitfall: raw `Value = 0` checks.** IEEE 754 `NaN` and `Infinity` values have specific bit patterns, but any code comparing `Value = 0` should also guard against negative zero. The `IsActualZero` helper in the arithmetic evaluator encapsulates this: `(Value = 0) and not IsNaN and not IsInfinite`. Similarly, the `NumericRank` helper in `Goccia.Values.ArrayValue.pas` maps each special value to a distinct sort key.
+
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/value-system.md]: Fixed in 1c9c9982. Clarified that `IsActualZero` intentionally treats `-0` as zero (correct for exponentiation where `(-0)^0 === 1`), and noted that contexts needing `-0` distinction should use `IsNegativeZero` explicitly.

File: docs/errors.md [hunk 20]
```diff
@@ -0,0 +1,375 @@
...
+- **Error types** -- `Error`, `TypeError`, `ReferenceError`, `RangeError`, `SyntaxError`, `URIError`, `AggregateError`, `SuppressedError`, plus `TimeoutError` for the `--timeout` flag
+- **Parser errors** -- Displayed with source context, a caret pointing to the exact column, and optional suggestion text (e.g., "Use 'let' or 'const' instead")
+- **Runtime errors** -- Carry `name`, `message`, `stack`, and optional `cause`; catchable with `try`/`catch`/`finally`
+- **JSON output** -- `--output=json` wraps every execution result in a structured envelope with `ok`, `error.type`, `error.message`, `error.line`, and `error.column`
+- **`Error.cause`** -- All error constructors accept an options bag with a `cause` property for error chaining (ES2022+)
+
+## Error Types
+
+GocciaScript supports the standard ECMAScript error constructors plus two additional types. All error types inherit from `Error` and work with `instanceof`.
+
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/errors.md]: Fixed in 1dd2163e. Added explicit note that `TimeoutError` is CLI-only and not exposed as a JavaScript constructor.

File: docs/errors.md [hunk 21]
```diff
@@ -0,0 +1,375 @@
...
+    at middle (script.js:5:3)
+    at outer (script.js:8:3)
+```
+
+Each frame shows the function name (or `<anonymous>`), the file path, and the line and column number. Frames are listed from innermost (most recent) to outermost.
+
+## Error.cause
+
+All error constructors accept an optional second argument -- an options object with a `cause` property. This follows [ES2022 Error Cause](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause):
+
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/errors.md]: Fixed in 1dd2163e. Clarified that the options argument position varies: second for `Error`/`TypeError`/`RangeError`/`ReferenceError`/`SyntaxError`/`URIError`, third for `AggregateError` (`new AggregateError(errors, message, options)`), fourth for `SuppressedError` (`new SuppressedError(error, suppressed, message, options)`). Note: GocciaScript's `SuppressedError` does support `cause` via a 4th argument — verified in `Goccia.Builtins.Globals.pas` (`AArgs.GetElement(3)`).

File: docs/testing.md [hunk 22]
```diff
@@ -203,9 +229,15 @@ test("JSON.stringify preserves round-trip precision for large fractional doubles
...
 
 ```bash
+# Interpreted mode (default)
 ./build/TestRunner tests
+
+# Bytecode mode
+./build/TestRunner tests --mode=bytecode
 ```
 
+Both modes must pass. CI runs the full suite in both interpreted and bytecode mode as separate matrix jobs.
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/testing.md]: Fixed in 1dd2163e. Added `--asi` to both interpreted and bytecode mode examples in the Run All Tests section, matching CI behavior.

File: docs/testing.md [hunk 23]
```diff
@@ -780,9 +382,31 @@ The `TestRunner` program:
...
+| Timeout handling | `--timeout` produces `TimeoutError` in JSON output |
+| Global injection | `--global`, `--globals` file/module injection, collision detection |
+| Stdin smoke tests | Piped input executes correctly in both interpreted and bytecode modes |
+| BenchmarkRunner output | `--format=json` produces valid JSON with benchmark structure |
+
+These tests do not require building from source -- they download pre-built binaries from the build job. To add a new CLI behavior check, add a step to the appropriate job in `pr.yml`.
+
+## Pascal Unit Tests (Tertiary)
+
+Pascal unit tests are the tertiary testing layer for low-level value system internals. They validate things that are difficult to test through JavaScript alone (e.g., memory layout, internal type conversion edge cases).
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/testing.md]: Fixed in 546f0d02. Broadened from "value system internals" to "low-level internals that are difficult to validate through JavaScript alone (e.g., value representation, function/scope internals, and test-framework state behavior)" — matching the actual scope shown in the table.

File: docs/testing.md [hunk 24]
```diff
@@ -767,6 +356,19 @@ promise.catch((err) => {
 });
 ```
 
+## Testing Strategy Design Decision
+
+JavaScript end-to-end tests are the **primary** testing mechanism. Every new feature or bug fix must include tests that validate the behavior through the full pipeline (lexer → parser → evaluator) using the most public surface available.
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/testing.md]: Fixed in af5a7697. Changed "lexer → parser → evaluator" to "lexer → parser → interpreter/VM execution" and added the CI dual-mode note. This was missed earlier because the agent only updated the layer description at line 14 and didn't search for other pipeline references in the same file.

File: docs/contributing/workflow.md [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+Formatting, editor integration, and CI behavior are covered in [Tooling](tooling.md).
+
+## Feature workflow
+
+Every change should follow this sequence:
+
+1. **Create a branch** from `main` with a descriptive name (for example `feature/string-prototype-repeat`, `fix/nan-comparison`).
+2. **Implement** on that branch, following [Critical rules](../../CONTRIBUTING.md#critical-rules) and [Code style](code-style.md).
+3. **Annotate spec references** — For ECMAScript behavior, add `// ESYYYY` spec comments as described in [ECMAScript spec annotations](code-style.md#ecmascript-spec-annotations).
+4. **Add or update tests** — JavaScript tests under `tests/` are primary; Pascal units under `units/*.Test.pas` when you touch AST, evaluator, or value types. See [testing.md](../testing.md) and [Critical rules](../../CONTRIBUTING.md#critical-rules).
```
[the remark this answers is not in the document]
Comment (frostney) [reply docs/contributing/workflow.md:55]: Not applicable — there are no dedicated scope chain Pascal unit tests. The Pascal test files that reference 'Scope' (`Compiler.Test.pas`, `TestAssertions.Test.pas`, `ScriptLoader.Globals.Test.pas`) test other concerns; scope chain behavior is tested through JavaScript E2E tests. Adding 'scope chain' to the trigger list would suggest Pascal tests exist for it when they don't.

File: scripts/check-doc-duplication.ts [hunk 26]
```diff
@@ -0,0 +1,758 @@
...
+const MIN_WORDS = parseInt(process.env.DOC_DUP_MIN_WORDS ?? "35", 10);
+const MIN_LOCATIONS = parseInt(process.env.DOC_DUP_MIN_LOCATIONS ?? "2", 10);
+const FUZZY_THRESHOLD = parseFloat(process.env.DOC_DUP_FUZZY_THRESHOLD ?? "0.65");
+const FUZZY_MIN_WORDS = parseInt(process.env.DOC_DUP_FUZZY_MIN_WORDS ?? "15", 10);
+const NUM_HASHES = 128;
+const NUM_BANDS = 32;
+const ROWS_PER_BAND = NUM_HASHES / NUM_BANDS; // 4
+const EXTENSIONS = new Set([".md", ".mdx"]);
+const IGNORE_DIRS = new Set(["node_modules", ".git", "dist", "build", ".next", "vendor"]);
+
```
[the remark this answers is not in the document]
Comment (frostney) [reply scripts/check-doc-duplication.ts:123]: Fixed — added `.agents` to `IGNORE_DIRS` so vendored skill docs are excluded from the scan.

File: scripts/check-doc-duplication.ts [hunk 27]
```diff
@@ -0,0 +1,758 @@
...
+    const slug = findSectionSlug(headings, canon.line);
+    const relPath = relative(root, canon.file);
+    const anchor = slug ? `#${slug}` : "";
+    const label = canon.section ?? basename(canon.file, ".md");
+    suggestions.push({
+      type: "exact",
+      canonical: { file: canon.file, line: canon.line, section: canon.section, slug },
+      duplicates: sorted.slice(1),
+      linkText: `See [${label}](./${relPath}${anchor})`,
+    });
```
[the remark this answers is not in the document]
Comment (frostney) [reply scripts/check-doc-duplication.ts:567]: Acknowledged — `--suggest` mode is opt-in and not used in CI. The current root-relative links are sufficient for the primary use case (human reading the console output). Per-duplicate relative paths would add complexity for a rarely-used feature; skipping for now.

File: scripts/check-doc-duplication.ts [hunk 28]
```diff
@@ -0,0 +1,758 @@
...
+
+    const lines = content.split("\n");
+    const cleaned: string[] = [];
+    let inCode = false;
+    for (const line of lines) {
+      if (/^```/.test(line.trim())) { inCode = !inCode; cleaned.push(""); continue; }
+      cleaned.push(inCode ? "" : line);
+    }
+    corpus.push(...tokenise(cleaned.join("\n"), file));
+    allParagraphs.push(...extractParagraphs(content, file, headings));
```
[the remark this answers is not in the document]
Comment (frostney) [reply scripts/check-doc-duplication.ts:729]: Fixed — a unique sentinel word (`\0FILE_BOUNDARY_N`) is now inserted between files in the corpus, so the suffix array cannot match phrases that span across file boundaries.
